spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -0,0 +1,181 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Entity.Builder;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.Query;
+import com.google.common.annotations.VisibleForTesting;
+
+import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreEntityConverter;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.data.mapping.PersistentProperty;
+import org.springframework.util.Assert;
+
+/**
+ * An implementation of {@link DatastoreOperations}.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreTemplate implements DatastoreOperations {
+
+	private final Datastore datastore;
+
+	private final DatastoreEntityConverter datastoreEntityConverter;
+
+	private final DatastoreMappingContext datastoreMappingContext;
+
+	public DatastoreTemplate(Datastore datastore,
+			DatastoreEntityConverter datastoreEntityConverter,
+			DatastoreMappingContext datastoreMappingContext) {
+		Assert.notNull(datastore, "A non-null Datastore service object is required.");
+		Assert.notNull(datastoreEntityConverter,
+				"A non-null DatastoreEntityConverter is required.");
+		Assert.notNull(datastoreMappingContext,
+				"A non-null DatastoreMappingContext is required.");
+		this.datastore = datastore;
+		this.datastoreEntityConverter = datastoreEntityConverter;
+		this.datastoreMappingContext = datastoreMappingContext;
+	}
+
+	@Override
+	public <T> T findById(Object id, Class<T> entityClass) {
```

Comment:
Since the key factory seems to only support Strings and longs, I think it might be more user-friendly to overload this method with `findById(String id, ...)` and `findById(long id, ...)`. We can also have `findByKey(Key key,...)`.

Reply:
Doing this would preclude the use of the converters that handle this for the user. Also, the examples I followed, the JDBC template and operations and mongo template and operations, support Object id type. 

I think overloading it would shift the responsibility of this conversion to the user when it's not necessary because we will have access to the converters in this template eventually anyway.

Reply:
This may be an instance where we deviate from `MongoTemplate` if it makes sense. How would it preclude the use of custom converters? Any method in the template can still call a converter?

Reply:
I didn't mean we couldn't use the custom converters anywhere else in the class, but only that we wouldn't be able to use them for the ID conversion when the user wants to do a fast convenient find operation. For example, let's say he has a Car object and a converter of Car to its ID string that isn't just a trivial getter method (perhaps it synthesizes it from several attributes of the car). He wouldn't be able to find by the Car object. 

@mp911de @olivergierke what do you think?

Reply:
From an API perspective, the Template API comes already with a lot of methods so adding overloads increases the API surface. 

Adding overloads would triple the number of methods that accept an Id. I propose to stick with `Object id` which is a common approach and it allows for extensibility in the future (Custom conversions, additional Id types).